creusot-contracts-proc/src/lib.rs

```diff
@@ -406,14 +405,17 @@ pub fn proof_assert(assertion: TS1) -> TS1 {
 }
 
 #[proc_macro]
-pub fn ghost(assertion: TS1) -> TS1 {
-    let assertion = TokenStream::from(assertion);
+pub fn gh(assertion: TS1) -> TS1 {
+    let assert = parse_macro_input!(assertion as Assertion);
+    let assert_body = pretyping::encode_block(&assert.0).unwrap();
+
     TS1::from(quote! {
         {
-            ::creusot_contracts::ghost::Ghost::from_fn(
+            ::creusot_contracts::__stubs::ghost_from_fn(
                 #[creusot::no_translate]
                 #[creusot::spec]
-                #[creusot::spec::ghost] || { ::creusot_contracts::ghost::Ghost::new (#assertion) }
+                #[creusot::spec::ghost]
+                || { ::creusot_contracts::ghost::Ghost::new (#assert_body) }
             )
```

Comment:
its worth doing another pass and seeing if we can't just do the same thing as `logic` though I suspect not due to interactions with initialization checking etc,..

Reply:
What do you want to do?

Reply:
I'll take a stab at this later, I think I have the most expertise fighting `rustc` on this aspect.